source/common/stream_info/filter_state_impl.cc

```diff
@@ -152,5 +167,44 @@ void FilterStateImpl::maybeCreateParent(ParentAccessMode parent_access_mode) {
   parent_ = lazy_create_ancestor.first;
 }
 
+absl::Status FilterStateImpl::addAncestor(const FilterStateSharedPtr& ancestor) {
+  if (!ancestor) {
+    return absl::OkStatus();
+  }
+  if (ancestor->lifeSpan() <= life_span_) {
+    return absl::FailedPreconditionError(absl::StrCat("Ancestor lifespan ", ancestor->lifeSpan(),
+                                                      " must be larger than current lifespan ",
+                                                      life_span_));
+  }
+  if (parent_) {
+    if (parent_.get() == ancestor.get()) {
+      return absl::OkStatus();
+    }
+    return parent_->addAncestor(ancestor);
+  }
+  if (auto* ancestor_ptr = absl::get_if<FilterStateSharedPtr>(&ancestor_);
+      ancestor_ptr && *ancestor_ptr) {
+    if (ancestor_ptr->get() == ancestor.get()) {
+      return absl::OkStatus();
+    }
+    return (*ancestor_ptr)->addAncestor(ancestor);
+  }
+  if (auto* lazy_create_ancestor = absl::get_if<LazyCreateAncestor>(&ancestor_);
+      lazy_create_ancestor && lazy_create_ancestor->first) {
+    if (lazy_create_ancestor->first.get() == ancestor.get()) {
+      return absl::OkStatus();
+    }
+    return lazy_create_ancestor->first->addAncestor(ancestor);
+  }
+  for (const auto& [key, _] : data_storage_) {
```

Comment:
I think this is not checking for conflicts two levels down? E.g. between filter chain (this) and connection (argument)?

Reply:
I actually missed a case. The filter state contract says that a key can exist at exactly one lifetime. If we have a key inside both `this` and the proposed ancestor, that's a problem, and we should fail.